Implemented resize functionality fix #100, #342

seytar opened the pull request:
<!-- EDIT TITLE PLEASE -->
<!-- It should be one of them
  <ISSUE TYPE>: Short Description (<CLOSING TYPE> #<ISSUE NUMBERS>)
  ex)
  feat: add new feature (close #111)
  fix: wrong behavior (fix #111)
  chore: change build tool (ref #111)
-->

<!-- SPECIFY A ISSUE TYPE AT HEAD
  feat: A new feature
  fix: A bug fix
  docs: Documentation only changes
  style: Changes that do not affect the meaning of the code (white-space, formatting etc)
  refactor: A code change that neither fixes a bug or adds a feature
  perf: A code change that improves performance
  test: Adding missing tests
  chore: Changes to the build process or auxiliary tools and libraries such as documentation generation
-->

<!-- ADD CLOSING TYPE AND ISSUE NUMBER AT TAIL
  (<CLOSING TYPE> #<ISSUE NUMBERS>)
  close: resolve not a bug(feature, docs, etc) completely
  fix: resolve a bug completely
  ref: not fully resolved or related to
-->

### Please check if the PR fulfills these requirements

- [x] It's submitted to right branch according to our branching model
- [x] It's right issue type on title
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)
- [x] It does not introduce a breaking change or has description for the breaking change

### Description

---

Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

lja1018 commented on apps/image-editor/src/css/submenu.styl:
```diff
@@ -62,11 +62,12 @@
             width: 55px;
             white-space: nowrap;
-    .{prefix}-submenu-align:first-child
+    .{prefix}-submenu-align:first-child, .{prefix}-submenu-align-mr-0
```
Where is the style used?
I can't find it.

seytar replied:
Ah, sorry. It don't used anywhere :)

seytar replied:
fixed.

lja1018 commented on apps/image-editor/src/js/command/resize.js:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author NHN. FE Development Team <dl_javascript@nhn.com>
+ * @fileoverview Rotate an image
```
Maybe `resize`...

lja1018 commented on apps/image-editor/src/js/component/resize.js:
```diff
@@ -0,0 +1,118 @@
+import { componentNames } from '@/consts';
+
+/**
+ * Zoom components
+ * @param {Graphics} graphics - Graphics instance
+ * @extends {Component}
+ * @class Zoom
```
`resize` components :)

lja1018 reviewed (commented):
Thank you for the contributing.
It's very cool feature :)
However, the icon can be modified because we have our designer.

Once again, thank you for the contributing!!

lja1018 commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +413,88 @@ export default {
+    return extend(
+      {
+        getCurrentDimensions: () => {
+          return this._graphics.getComponent(componentNames.RESIZE).getCurrentDimensions();
```
You are using the components of graphics.
Could you create API for graphics?
like this...
```js
return this._graphics.getCurrentDimensions();
```
The other parts are the same.

lja1018 commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +413,88 @@ export default {
+                dimensions.width = currentDimensions.width;
+              }
+              break;
+            default:
```
I think it would be good to specify the default dimensions value.

lja1018 commented on apps/image-editor/src/js/component/resize.js:
```diff
@@ -0,0 +1,118 @@
+   */
+  getCurrentDimensions() {
+    if (!this._dimensions) {
+      const canvasImage = this.getCanvasImage();
+      this._dimensions = {
+        width: canvasImage.width,
+        height: canvasImage.height,
```
I think it would be good to use shorthand.
Maybe...
```js
const { width, height } = this.getCanvasImage();
this._dimensions = { width, height };
```

lja1018 commented on apps/image-editor/src/js/component/resize.js:
```diff
@@ -0,0 +1,118 @@
+  resize(dimensions) {
+    const canvasImage = this.getCanvasImage();
+    const scaleValues = {
+      scaleX: dimensions.width ? dimensions.width / canvasImage.width : canvasImage.scaleX,
+      scaleY: dimensions.height ? dimensions.height / canvasImage.height : canvasImage.scaleY,
+    };
+
+    if (canvasImage.scaleX !== scaleValues.scaleX || canvasImage.scaleY !== scaleValues.scaleY) {
+      canvasImage.set(scaleValues).setCoords();
+
+      this._dimensions = {
+        width: canvasImage.width * canvasImage.scaleX,
+        height: canvasImage.height * canvasImage.scaleY,
+      };
```
Likewise, it would be good to use shorthand.
```js
const { width, height, scaleX, scaleY } = this.getCanvasImage();
const { width: dimensionsWidth: height: dimensionsHeight }  = dismensions;
const scales = {
  scaleX: dimensionsWidth ? dimensionsWidth / width : scaleX,
  scaleY: dimensionsHeight ? dimensionsHeight / height : scaleY,
};
...
```

lja1018 commented on apps/image-editor/src/js/graphics.js:
```diff
@@ -582,7 +584,14 @@ class Graphics {
+  }
+
+  adjustCanvasDimensionBase(canvasImage = null) {
+    if (canvasImage === null) {
```
I think you can use 
```
if (!canvasImage)
```

lja1018 reviewed (commented):
Thank you for contributing. :)

I have two questions.
First of all, one test is being broken(`npm run test` command). Can you fix it? And could you add a test for `resize`?
Second, what would be the copyright on the icon you added? If there is no copyright issue, I think I will use it as it is.

jungeun-cho commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +406,84 @@ export default {
+          }
+        },
+        resize: (dimensions = null) => {
+          // const resizeComponent = this._graphics.getComponent(componentNames.RESIZE);
```
Please delete unnecessary comments.

jungeun-cho commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +406,84 @@ export default {
+            ['catch']((message) => Promise.reject(message));
+        },
+        reset: (standByMode = false) => {
+          // const resizeComponent = this._graphics.getComponent(componentNames.RESIZE);
```
Please delete unnecessary comments. too.

jungeun-cho commented on apps/image-editor/src/js/action.js:
```diff
@@ -541,7 +620,7 @@ export default {
         if (this.ui.submenu === 'text') {
           this.changeCursor('text');
-        } else if (this.ui.submenu !== 'draw' && this.ui.submenu !== 'crop') {
+        } else if (this.ui.submenu !== 'draw' && this.ui.submenu !== 'crop' && this.ui.submenu !== 'resize') {
```
How about changing the `if` conditional statement like this?
```js
else if (!['draw', 'crop', 'resize'].includes(submenu)) {
```

jungeun-cho commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +406,83 @@ export default {
+          switch(actor) {
+            case 'width':
+              dimensions.width = value;
+              if (lockState) {
+                dimensions.height = value / calcAspectRatio();
+              } else {
+                dimensions.height = currentDimensions.height;
+              }
+              break;
+            case 'height':
+              dimensions.height = value;
+              if (lockState) {
+                dimensions.width = value * calcAspectRatio();
+              } else {
+                dimensions.width = currentDimensions.width;
+              }
```
When lockState is turned on, validation should be performed when the size is calculated.

This is because it exceeds 4088.

seytar replied:
Added validation

jungeun-cho replied:
I checked validation.
However, When operated in conjunction with the zoom operation, a side effect occurs.

<img width="568" alt="스크린샷 2021-04-15 21 56 46" src="https://user-images.githubusercontent.com/43128697/114872983-caee4400-9e35-11eb-837c-e295a7e9d2e0.png">

@lja1018 Please leave helpful feedback regarding the zoom action.

lja1018 replied:
@seytar
It can be resolved by invoking an API that resets zoom prior to resize image.

jungeun-cho reviewed (commented):
Thank you for your contribution :) 👍

seytar commented:
> Thank you for contributing. :)
> 
> I have two questions.
> First of all, one test is being broken(`npm run test` command). Can you fix it? And could you add a test for `resize`?
> Second, what would be the copyright on the icon you added? If there is no copyright issue, I think I will use it as it is.

- I fixed broken test.
- I used that icon: [Flaticon Resize](https://www.flaticon.com/free-icon/maximize_633681?term=resize&page=1&position=22&page=1&position=22&related_id=633681&origin=tag) and it says "Attribution is required.".
- I can add `resize` tests

seytar commented:
> > Thank you for contributing. :)
> > I have two questions.
> > First of all, one test is being broken(`npm run test` command). Can you fix it? And could you add a test for `resize`?
> > Second, what would be the copyright on the icon you added? If there is no copyright issue, I think I will use it as it is.
> 
> * I fixed broken test.
> * I used that icon: [Flaticon Resize](https://www.flaticon.com/free-icon/maximize_633681?term=resize&page=1&position=22&page=1&position=22&related_id=633681&origin=tag) and it says "Attribution is required.".
> * I can add `resize` tests

Added all tests @lja1018

lja1018 commented on apps/image-editor/src/js/action.js:
```diff
@@ -405,6 +406,109 @@ export default {
+            if (width > height) {
+              const pMax = max / aspectRatio;
+              const pMin = min * aspectRatio;
+              this.ui.resize.setMaxHeightValue(pMax < max ? pMax : max);
+              this.ui.resize.setMinWidthValue(pMin > min ? pMin : min);
+              this.ui.resize.setMaxWidthValue(max);
+              this.ui.resize.setMinHeightValue(min);
+            } else {
+              const pMax = max * aspectRatio;
+              const pMin = min / aspectRatio;
+              this.ui.resize.setMaxWidthValue(pMax < max ? pMax : max);
+              this.ui.resize.setMinHeightValue(pMin > min ? pMin : min);
+              this.ui.resize.setMaxHeightValue(max);
+              this.ui.resize.setMinWidthValue(min);
+            }
+          } else {
+            this.ui.resize.setMaxHeightValue(max);
+            this.ui.resize.setMaxWidthValue(max);
+            this.ui.resize.setMinWidthValue(min);
+            this.ui.resize.setMinHeightValue(min);
```
How about combining 4 functions that set `min` and `max` values? If you don't have to be called alone, you'd better put it together.
```js
// example
this.ui.resize.setLimit({
  minWidth: min,
  minHeight: min,
  maxWidth: max,
  maxHeight: max,
});
```
If you have a chance to be called alone, you can use it as it is. :)

lja1018 commented on apps/image-editor/src/js/component/resize.js:
```diff
@@ -0,0 +1,103 @@
+  getCurrentDimensions() {
+    const canvasImage = this.getCanvasImage();
+    if (!this._dimensions && canvasImage) {
+      const { width, height } = canvasImage;
+      this._dimensions = { width, height };
```
I think it's okay to use it in a single line.
`this._dimensions = canvasImage;`

seytar replied:
Yes, with this change will get same results. But _dimensions property should has just two property: width and height, and I think canvasImage is big object for this using.

lja1018 commented on apps/image-editor/tests/resize.spec.js:
```diff
@@ -0,0 +1,104 @@
+    expect(originalDimensions).toEqual(resizeModule.getCurrentDimensions());
+  });
+
+  it('"start()" should setted original dimensions', () => {
```
`should set` :)

lja1018 reviewed (approved):
Thank yor for your contribution. 👍 👍 👍 

I have some questions.
1. Should I add the image source to `README`?
2. Please add 1 line to `component/resize`(`import { Promise } from '@/util';`). Required for old browser support :)

seytar commented:
> Thank yor for your contribution.   
> 
> I have some questions.
> 
> 1. Should I add the image source to `README`?
> 2. Please add 1 line to `component/resize`(`import { Promise } from '@/util';`). Required for old browser support :)

1. I think yes.
2. Ok, added.

lja1018 commented:
@seytar 
Thank you for applying on all the reviews. :)
I will release it by next week at the latest.
